### PR TITLE
Add SoftwareSerial to HAL_STM32

### PR DESCRIPTION
**Pull request by sjasonsmith**
### Description

Add a SoftwareSerial implementation to HAL_STM32.
This is derived from the version in  Arduino_Core_STM32 1.8 (not yet released). It is modified to allow setting the interrupt property in Arduino_Core_STM32 1.7.

### Benefits

Allows migrating to STM32 5.7, without waiting for Arduino_Core_STM32 to release.

### CODE REVIEW

**Comment by sjasonsmith** [comment]
The interface to set the serial interrupt priority matches what it is in the Arduino_Core_STM32. This will simplify switching to the core version of SoftwareSerial when it is released.

**Comment by LinoBarreca** [comment]
we don't need to wait for their release, actually.
we can just modify platformio.ini to get the git of 1.8.0 instead of the package.
I'm working on it.

**Comment by LinoBarreca** [comment]
(have a look here: https://docs.platformio.org/en/latest/projectconf/section_env_platform.html#id2 )

**Comment by LinoBarreca** [comment]
but that crap doesn't work.....

**Comment by sjasonsmith** [comment]
I forgot to remove that SS_TIMER define for the skr pro. That is no longer needed, since SoftwareSerialM is gone.

**Comment by LinoBarreca** [comment]
@sjasonsmith did you find where the problem lied?

File: Marlin/src/HAL/HAL_STM32/timers.h [hunk 1]
```diff
@@ -33,6 +33,10 @@
 #define hal_timer_t uint32_t
 #define HAL_TIMER_TYPE_MAX 0xFFFFFFFF // Timers can be 16 or 32 bit
 
+#define SWSERIAL_TIMER_PRIORITY 0
```
Comment (LinoBarreca) [inline Marlin/src/HAL/HAL_STM32/timers.h:36]: timer priority 0 is too high. that's equal to systick. nothing should have that priority.
Increase this and stepper by one.

File: Marlin/src/HAL/HAL_STM32/timers.h [hunk 2]
```diff
@@ -69,6 +73,7 @@
     #define TEMP_TIMER 14
   #endif
 
+  // TIM7 is consumed by Software Serial if used.
```
Comment (LinoBarreca) [inline Marlin/src/HAL/HAL_STM32/timers.h:76]: Where is this specified? is it the default?

Comment (sjasonsmith) [reply Marlin/src/HAL/HAL_STM32/timers.h:76]: That is the default. It is probably safer to set it explicitly, but you have to do it in the variant file or with -DTIMER_SERIAL=TIMx.

File: Marlin/src/HAL/HAL_STM32/timers.h (hunk 1 above)
Comment (sjasonsmith) [reply Marlin/src/HAL/HAL_STM32/timers.h:36]: I had used that because it was what was used before (by 3rd-party SoftwareSerial code), but it can change.

**Comment by sjasonsmith** [comment]
@LinoBarreca, thanks for the changes! I agree with all the changes you made.

**Pull request merged**